sysfs/class_drm.go

```diff
@@ -0,0 +1,124 @@
+	"regexp"
+	"strconv"
+	"syscall"
+
+	"github.com/prometheus/procfs/internal/util"
+)
+
+var (
+	// list of supported device drivers.
+	drmDeviceDrivers = []string{"amdgpu"}
+)
+
+// ClassDrmCardStats contains info from files in /sys/class/drm/card<card>/device
+// for a single card.
+// https://www.kernel.org/doc/html/latest/gpu/drivers.html
+// https://www.kernel.org/doc/html/latest/gpu/amdgpu.html
+type ClassDrmCardStats struct {
+	Name                          string // The card name.
+	Driver                        string // The kernel driver loaded for this card.
+	GpuBusyPercent                uint64 `amdgpu:"gpu_busy_percent"`                  // How busy the GPU is as a percentage.
+	MemoryGttSize                 uint64 `amdgpu:"mem_info_gtt_total"`                // The size of the graphics translation table (GTT) block in bytes.
+	MemoryGttUsed                 uint64 `amdgpu:"mem_info_gtt_used"`                 // The used amount of the graphics translation table (GTT) block in bytes.
+	MemoryVisibleVramSize         uint64 `amdgpu:"mem_info_vis_vram_total"`           // The size of visible VRAM in bytes.
+	MemoryVisibleVramUsed         uint64 `amdgpu:"mem_info_vis_vram_used"`            // The used amount of visible VRAM in bytes.
+	MemoryVramSize                uint64 `amdgpu:"mem_info_vram_total"`               // The size of VRAM in bytes.
+	MemoryVramUsed                uint64 `amdgpu:"mem_info_vram_used"`                // The used amount of VRAM in bytes.
+	MemoryVramVendor              string `amdgpu:"mem_info_vram_vendor"`              // The VRAM vendor name.
+	PowerDpmForcePerformanceLevel string `amdgpu:"power_dpm_force_performance_level"` // The current power performance level
+	UniqueID                      string `amdgpu:"unique_id"`                         // The unique ID of the GPU that will persist from machine to machine.
+}
+
+// ClassDrmCardStats returns DRM card metrics for all cards.
+func (fs FS) ClassDrmCardStats() ([]ClassDrmCardStats, error) {
+	cards, err := filepath.Glob(fs.sys.Path("class/drm/card[0-9]"))
+	if err != nil {
+		return nil, err
+	}
+
+	stats := make([]ClassDrmCardStats, 0, len(cards))
+	for _, card := range cards {
+		cardStats, err := parseClassDrmCard(card)
+		if err != nil {
+			if errors.Is(err, syscall.ENODATA) {
+				continue
+			}
+			return nil, err
+		}
+		cardStats.Name = filepath.Base(card)
+		stats = append(stats, cardStats)
+	}
+	return stats, nil
+}
+
+func parseClassDrmCard(card string) (ClassDrmCardStats, error) {
+	stats := ClassDrmCardStats{Name: card}
+	uevent, err := util.SysReadFile(filepath.Join(card, "device/uevent"))
+	if err != nil {
+		return stats, err
+	}
+
+	for _, driver := range drmDeviceDrivers {
+		match, err := regexp.MatchString(fmt.Sprintf("DRIVER=%s", driver), uevent)
+		if err != nil {
+			return stats, err
+		}
+		if match {
+			stats.Driver = driver
+			break
+		}
+	}
+
+	// Use struct tags to dynamically read and parse values depending on the driver.
```

Comment:
So to support other drivers, we'd add a new tag to an existing field to read from a different sysfs field?
So far we tried to avoid using reflect where possible. But seems like a reasonable pattern, just that if we want to go that route, we should do it this way for similar cases as well. @pgier @SuperQ wdyt?

Reply:
Correct, I came up with this solution to avoid adding custom function per driver.
But if using `reflect` is an issue, I can drop it.

Reply:
I think while it's a smart approach and yields a condensed mapping in the struct above, as someone who is not really familiar with how the reflection API works, the code below is a bit hard for me to understand. It can also be surprising to readers because it's not an established reflection use case pattern in Go. So personally I'd prefer a bit more verbosity for improved clarity / simplicity - the Go way :)

Reply:
Maybe we can create a package similar to https://github.com/Netflix/go-env but for mapping `procfs` and `sysfs` entries to struct fields later.

Reply:
If we go down that route, we'd need something that works for other parts of procfs.
So I'd say you can either change this to use a simple amd specific struct or propose a struct tagged based option with examples how it would look like for other parts of procfs. The later is more involved and there will be probably quite some back and forth..

Reply:
I spent sometime last month and I have 3 proposals:
1. Use struct tags and reflection (what we discussed above but at a generic lib)
2. Use interfaces by defining `Class`, `Device`, etc. and a library which accepts those interfaces. Then we can define a `GPU` structure which satisfies the interfaces methods. The library handles reading and generic parsing while the `GPU` structure handles any adjustments or conversions needed on read values, etc.
3. Use code generation, a command will be added to read a `proc` or `sysfs` path which will then generate all the structures and methods for reading and accessing the read values, then custom code can be added on top to handle adjustments or conversions needed on the read values.

Sorry for the late response and the long comment.
Let me know what you think is the proper way to proceed.

Reply:
2. sounds like the interfaces might be a bit to abstract to be useful. I think I'd prefer 1 but not convinced it's worth it over just using specific hardcoded structs, at least until we have the need to support multiple variants.

Reply:
I'll start working on 2 then.

Reply:
I misread your comment yesterday so ignore my deleted comment, and indeed interfaces make it abstract.
I'll rewrite the PR to use static way already used in the project.